pennylane/transforms/op_transforms.py

```diff
@@ -0,0 +1,481 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the @op_transform decorator.
+"""
+# pylint: disable=protected-access
+import functools
+import inspect
+import os
+import warnings
+
+import pennylane as qml
+
+
+class OperationTransformError(Exception):
+    """Raised when there is an error with op_transform logic"""
+
+
+def _make_tape(obj, wire_order, *args, **kwargs):
+    """Given an input object, which may be:
+
+    - an object such as a tape or a operation, or
+    - a callable such as a QNode or a quantum function
+      (alongside the callable arguments ``args`` and ``kwargs``),
+
+    this function constructs and returns the tape/operation
+    represented by the object.
+
+    The ``wire_order`` argument determines whether a custom wire ordering
+    should be used. It not provided, the wire ordering defaults to the
+    objects wire ordering accessed via ``obj.wires``.
+    """
+    if isinstance(obj, qml.QNode):
+        # user passed a QNode, get the tape
+        obj.construct(args, kwargs)
+        tape = obj.qtape
+        wires = obj.device.wires
+
+    elif isinstance(obj, qml.tape.QuantumTape):
+        # user passed a tape
+        tape = obj
+        wires = tape.wires
+
+    elif inspect.isclass(obj) and issubclass(obj, qml.operation.Operator):
+        with qml.tape.stop_recording():
+            tape = obj(*args, **kwargs)
+
+        wires = tape.wires
+
+    elif callable(obj):
+        # user passed something that is callable but not a tape or QNode.
+        tape = qml.transforms.make_tape(obj)(*args, **kwargs)
+        wires = tape.wires
+
+        # raise exception if it is not a quantum function
+        if len(tape.operations) == 0 and len(tape.measurements) == 0:
+            raise OperationTransformError("Quantum function contains no quantum operations")
+
+    # if no wire ordering is specified, take wire list from tape/device
+    wire_order = wires if wire_order is None else qml.wires.Wires(wire_order)
+
+    # check that all wire labels in the circuit are contained in wire_order
+    if not set(tape.wires).issubset(wire_order):
+        raise OperationTransformError(
+            f"Wires in circuit {tape.wires.tolist()} are inconsistent with "
+            f"those in wire_order {wire_order.tolist()}"
+        )

```

Comment:
This may be expensive, but probably not wise to just ignore, right?

Reply:
Ah, I copied this from `get_unitary_matrix`, what do you think? Worth removing?

Reply:
I really don't know...maybe leave for now, and sacrifice for speed later!